core/src/main/java/fj/data/Eval.java

```diff
@@ -0,0 +1,209 @@
+  }
+
+  /**
+   * Transforms <code>Eval<A></code> into a <code>Eval<B></code> using
+   * the given function that directly produces <code>Eval<B></code>.
+   *
+   * Note: the computation of the given transformation is always lazy,
+   * even if it invoked for an eager <code>Now</code> instance. This computation
+   * is performed in O(1) stack space.
+   *
+   * @param f the transformation function.
+   * @return a transformed evaluation.
+   */
+  public <B> Eval<B> bind(final F<A, Eval<B>> f) {
+    return bindTrampoline(f, asTrampoline());
+  }
+
+  /**
+   * Transforms the current instance into a trampoline instance.
+   */
+  TrampolineEval<A> asTrampoline() {
+    return pureTrampoline(this);
+  }
+
+  /**
+   * Represents an eager computation.
+   */
+  private static final class Now<A> extends Eval<A> {
+    private final A a;
+
+    private Now(A a) {
+      this.a = a;
+    }
+
+    @Override
+    public A value() {
+      return a;
+    }
+  }
+
+  /**
+   * Represents a lazy computation that is evaluated only once.
+   */
+  private static final class Later<A> extends Eval<A> {
+    private final P1<A> memo;
+
+    private Later(F0<A> producer) {
+      this.memo = P.hardMemo(producer);
+    }
+
+    @Override
+    public A value() {
+      return memo._1();
+    }
+  }
+
+  /**
+   * Represents a lazy computation that is evaluated every time when it's requested.
+   */
+  private static final class Always<A> extends Eval<A> {
+    private final F0<A> supplier;
+
+    private Always(F0<A> supplier) {
+      this.supplier = supplier;
+    }
+
+    @Override
+    public A value() {
+      return supplier != null ? supplier.f() : null;
+    }
+  }
+
+  /**
+   * A helper abstraction that allows to perform recursive lazy transformations in O(1) stack space.
+   */
+  private static abstract class TrampolineEval<A> extends Eval<A> {
+
+    protected abstract Trampoline<A> trampoline();
+
+    @Override
+    public A value() {
+      return trampoline().run();
+    }
+
+    @Override
+    TrampolineEval<A> asTrampoline() {
+      return this;
+    }
+  }
+
+  private static final class PureTrampolineEval<A> extends TrampolineEval<A> {
+    private final Eval<A> start;
+
+    private PureTrampolineEval(Eval<A> start) {
+      this.start = start;
+    }
+
+    @Override
+    protected Trampoline<A> trampoline() {
+      return Trampoline.suspend(P.lazy(() -> Trampoline.pure(start.value())));
+    }
+  }
+
+  private static final class BindTrampolineEval<A, B> extends TrampolineEval<B> {
+    private final TrampolineEval<A> next;
+    private final F<A, Eval<B>> f;
+
+    private BindTrampolineEval(F<A, Eval<B>> f, TrampolineEval<A> next) {
+      this.next = next;
+      this.f = f;
+    }
+
+    @Override
+    protected Trampoline<B> trampoline() {
+      return Trampoline.suspend(P.lazy(() -> next.trampoline().map(v -> f.f(v).value())));
+    }
+  }
+
+  private static <A> TrampolineEval<A> pureTrampoline(Eval<A> start) {
```

Comment:
not sure this private methods (`pureTrampoline` and `bindTrampoline`) add much. could you inline them?